### ISSUE TITLE
Problems related to pretraining Bert

### ISSUE THREAD

**Issue by sw778** [lulutang0608/Point-BERT#51]
Hi! When I pretrain Bert, I get the following error on the first round of validation every epoch I train. I want to train for `3000` epochs, when I set` --val_freq 3001`, there is no error, but I can't get the `Losses` for each epoch(as shown in the second picture). Do you have any good suggestions, is this related to memory leaks? thank you for your reply!

![图片1](https://user-images.githubusercontent.com/81848293/175803512-1bdf2b2e-738d-43cd-a067-c8143c40fa5b.png)

![175802719-486a924b-ec20-47bb-89c3-535bc129f1cd](https://user-images.githubusercontent.com/81848293/175803521-f1c0f7a6-5b22-4c3f-a2aa-f325d212d388.png)

**Comment by yuxumin** [comment]
Hi, sprry for the late reply.

I can't figure out the problem you are facing from fig1. (if you don't set `val_freq`, it may due to segmentation fault caused by the conflict between svm and pytorch)

> there is no error, but I can't get the Losses for each epoch

Sorry that i do not know what that means. I see the losses are printed out for each epoch training. And the warning you pointed out in fig2 does not matter. (You can change `True` to `false` in this line https://github.com/lulutang0608/Point-BERT/blob/master/tools/runner_BERT_pretrain.py#L81 to avoid it)

**Comment by Ma-Weijian** [comment]
Hi， I've met nearly the same issue.

As I've mentioned in #65 , I found the training of dvae too slow and I changed the training procedure of dvae into ddp.

However, the training crashes for every 27 epochs. During the training procedure, I've discovered that the memory usage has increased for several GBs after each validation. My hardware environment is 2x A100 40GB and about 130GB RAM.

From what I and @sw778 have discovered, I believe there must be some memory leakage in ddp pipeline used in this repo.

@yuxumin @sw778  Anything helps. a million thanks!

**Comment by Ma-Weijian** [comment]
@sw778 One more thing, could u plz tell me the size of your RAM when u met this issue?

**Comment by yuxumin** [comment]
@Ma-Weijian 
Hey there, sorry for the delay in getting back to you. Let's talk about issue 51. It actually pertains to the pretraining process of Point-BERT (specifically, Mask point modeling) and not dVAE.

The problem you're encountering seems to be related to the picture saved during validation. Take a look at this line (https://github.com/lulutang0608/Point-BERT/blob/master/tools/runner.py#L246). If you don't need to see the visualization results for the dVAE you trained, you can try commenting out this function.

Hope that helps! Let me know if you have any more questions.

**Comment by yuxumin** [comment]
Additionally, the slower inference time can be attributed to the batch size used during validation (bs=1). You can find the relevant line here (https://github.com/lulutang0608/Point-BERT/blob/master/main.py#L45). I initially set the batch size to 1 to address instability issues with chamfer distance calculation. When the batch size exceeds 1, the chamfer distance may become larger, which could be related to the precision of PyTorch. Unfortunately, I haven't been able to resolve this issue yet.

To adjust the batch size, you can make the following changes in the `main.py` file at line 45:

```python
# main.py L45
if args.distributed:
    assert config.total_bs % world_size == 0
    config.dataset.train.others.bs = config.dataset.val.others.bs = config.dataset.test.others.bs = config.total_bs // world_size
else:
    config.dataset.train.others.bs = config.dataset.val.others.bs = config.dataset.test.others.bs = config.total_bs
```

Feel free to try different batch sizes and observe their impact on the inference time (and chamfer distance calculation, but maybe unnecessary :).

**Comment by Ma-Weijian** [comment]
Thanks for the reply. I'll have a try.

But I still wonder why you choose to use data parallel when training dVAE. Since the framework of training dvae and masked point modeling are the same. Have you ever met something special that led to such choice?

**Comment by yuxumin** [comment]
Well, to be honest, I'm just feeling a bit lazy to rewrite it into DDP form :). After training dvae successfully in the very beginning of this work, I haven't actually used this piece of code again.
